node/src/accountant/db_access_objects/failed_payable_dao.rs

```diff
@@ -0,0 +1,778 @@
+    pub amount: u128,
+    pub timestamp: i64,
+    pub gas_price_wei: u128,
+    pub nonce: u64,
+    pub reason: FailureReason,
+    pub checked: u8,
+}
+
+pub enum FailureRetrieveCondition {
+    ByReason(FailureReason),
+}
+
+impl Display for FailureRetrieveCondition {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            FailureRetrieveCondition::ByReason(reason) => {
+                write!(f, "WHERE reason IS '{:?}'", reason)
+            }
+        }
+    }
+}
+
+pub trait FailedPayableDao {
+    fn get_tx_identifiers(&self, hashes: &HashSet<TxHash>) -> TxIdentifiers;
+    fn insert_new_records(&self, txs: &[FailedTx]) -> Result<(), FailedPayableDaoError>;
+    fn retrieve_txs(&self, condition: Option<FailureRetrieveCondition>) -> Vec<FailedTx>;
+    fn update_tx_failures(&self, hash_map: &FailureUpdates) -> Result<(), FailedPayableDaoError>;
+    fn delete_records(&self, hashes: &HashSet<TxHash>) -> Result<(), FailedPayableDaoError>;
+}
+
+#[derive(Debug)]
+pub struct FailedPayableDaoReal<'a> {
+    conn: Box<dyn ConnectionWrapper + 'a>,
+}
+
+impl<'a> FailedPayableDaoReal<'a> {
+    pub fn new(conn: Box<dyn ConnectionWrapper + 'a>) -> Self {
+        Self { conn }
+    }
+}
+
+impl FailedPayableDao for FailedPayableDaoReal<'_> {
+    fn get_tx_identifiers(&self, hashes: &HashSet<TxHash>) -> TxIdentifiers {
+        let hashes_vec: Vec<TxHash> = hashes.iter().copied().collect();
+        let sql = format!(
+            "SELECT tx_hash, rowid FROM failed_payable WHERE tx_hash IN ({})",
+            comma_joined_stringifiable(&hashes_vec, |hash| format!("'{:?}'", hash))
+        );
+
+        let mut stmt = self
+            .conn
+            .prepare(&sql)
+            .unwrap_or_else(|_| panic!("Failed to prepare SQL statement"));
+
+        stmt.query_map([], |row| {
+            let tx_hash_str: String = row.get(0).unwrap_or_default();
+            let tx_hash = TxHash::from_str(&tx_hash_str[2..]).unwrap_or_default();
+            let row_id: u64 = row.get(1).unwrap_or_default();
+
+            Ok((tx_hash, row_id))
+        })
+        .unwrap_or_else(|_| panic!("Failed to execute query"))
+        .filter_map(Result::ok)
+        .collect()
+    }
+
+    fn insert_new_records(&self, txs: &[FailedTx]) -> Result<(), FailedPayableDaoError> {
+        if txs.is_empty() {
+            return Err(FailedPayableDaoError::EmptyInput);
+        }
+
+        let unique_hashes: HashSet<TxHash> = txs.iter().map(|tx| tx.hash).collect();
+        if unique_hashes.len() != txs.len() {
+            return Err(FailedPayableDaoError::InvalidInput(
+                "Duplicate hashes found in the input".to_string(),
+            ));
+        }
+
+        if !self.get_tx_identifiers(&unique_hashes).is_empty() {
+            return Err(FailedPayableDaoError::InvalidInput(
+                "Input hash is already present in the database".to_string(),
+            ));
+        }
```

Comment:
This might be quite expensive. Remember that this is communication between different processes (the db manager). Cannot you simply ignore this preliminary check (which would run every time!) and let the SQL be executed. I guess the insertion would show at its result if you count the number of changed rows (that's the output of execute). I guess we don't have to be so precise here, do we? I think this will run fairly a lot and I don't want to include this check every time so.

Reply:
I want to explore the option of preventing data corruption by throwing the error earlier.